lib/route.js

```diff
@@ -208,117 +201,132 @@ function buildRouting (options) {
-        opts.errorHandler || this[kErrorHandler],
-        opts.bodyLimit,
-        opts.logLevel,
-        opts.logSerializers,
-        opts.attachValidation,
-        this[kReplySerializerDefault],
-        opts.schemaErrorFormatter || this[kSchemaErrorFormatter]
+        config
       )
 
-      const headRouteExists = router.find('HEAD', path) != null
-
-      try {
-        router.on(opts.method, opts.url, { constraints }, routeHandler, context)
-      } catch (err) {
-        done(err)
-        return
+      if (opts.version) {
+        warning.emit('FSTDEP006')
+        constraints.version = opts.version
       }
 
-      const { exposeHeadRoute } = opts
-      const hasRouteExposeHeadRouteFlag = exposeHeadRoute != null
-      const shouldExposeHead = hasRouteExposeHeadRouteFlag ? exposeHeadRoute : globalExposeHeadRoutes
+      const headRouteExists = opts.method === 'HEAD' && router.find('HEAD', url, constraints) != null
+
+      // TODO: improve validation for auto expose HEAD route before
+      // adding a new route
+
+      // There's a bug when ignoreTrailingSlash !== true
+      // See
+      /*
+      └── /prefix
+          └── / (GET)
+              / (HEAD)
+              └── / (GET)
+      **/
+
+      if (!headRouteExists) {
+        try {
+          router.on(opts.method, opts.url, { constraints }, routeHandler, context)
+        } catch (error) {
+          const isDuplicatedRoute = Array.isArray(error.message.match(`Method '${opts.method}' already declared for route '${opts.url}'`))
+          if (isDuplicatedRoute) {
+            throw new FST_ERR_DUPLICATED_ROUTE(opts.method, opts.url)
+          }
 
-      if (shouldExposeHead && options.method === 'GET' && !headRouteExists) {
-        const onSendHandlers = parseHeadOnSendHandlers(opts.onSend)
-        prepareRoute.call(this, 'HEAD', path, { ...opts, onSend: onSendHandlers })
-      } else if (headRouteExists && exposeHeadRoute) {
-        warning.emit('FSTDEP007')
+          throw error
+        }
       }
 
-      // It can happen that a user registers a plugin with some hooks *after*
-      // the route registration. To be sure to also load those hooks,
-      // we must listen for the avvio's preReady event, and update the context object accordingly.
-      avvio.once('preReady', () => {
-        for (const hook of lifecycleHooks) {
-          const toSet = this[kHooks][hook]
-            .concat(opts[hook] || [])
-            .map(h => {
-              const bound = h.bind(this)
-
-              // Track hooks deprecation markers
-              if (hook === 'preParsing') {
-                // Check for deprecation syntax
-                if (h.length === (h.constructor.name === 'AsyncFunction' ? 2 : 3)) {
-                  warning.emit('FSTDEP004')
-                  bound[kHooksDeprecatedPreParsing] = true
+      this.after((notHandledErr, done) => {
+        // Send context async
+        context.errorHandler = opts.errorHandler || this[kErrorHandler]
```

Comment:
Why set the context values here instead of `constructor`? [here](https://github.com/fastify/fastify/pull/2954/files?diff=unified&w=1#diff-cf6996cafd1df0bc168a5c5f3a7470f11fc3aea9f908091d5dd6be2e7b30a2c3L240)

Reply:
To be honest, I'm not 100% sure of the reason behind it. Maybe @mcollina knows more about the reason behind it as is based on [his suggestion](https://github.com/fastify/fastify/issues/2842#issuecomment-800962405).

Is also possible that I misunderstood the suggestion, if thats case, I'm open to change the implementation 🙂